### PR TITLE
chore: join owner, template, and org in new workspace view

### LINKED ISSUE #15114: RBAC error when viewing all users' workspaces
Given:
- an org
- a group with a member assigned (with only a "Member" role)
- a role with all permissions afforded and assigned to the member
- a template owned by an org
- a workspace using the org's template

When attempting to view all workspaces in an org as the above member, I encounter this error:
Image

This stems from:
https://github.com/coder/coder/blob/eb1602bba472f2cb4eff9c8cac58c5b69356052c/coderd/workspacebuilds.go#L667-L670

### PR DESCRIPTION
**Pull request by Emyrk**
Closes https://github.com/coder/coder/issues/15114

# What this does

Joins in fields like `username`, `avatar_url`, `organization_name`, `template_name` to `workspaces` via a **view**.

Views are a nuisance to maintain, however they have become the only clean way to handle RBAC related objects for:
- [Groups](https://github.com/coder/coder/blob/bbbe7aa2687671c8362ab3561c040d34ba0cca19/coderd/database/dump.sql#L687-L687)
- [Template Versions](https://github.com/coder/coder/blob/bbbe7aa2687671c8362ab3561c040d34ba0cca19/coderd/database/dump.sql#L1212-L1212)
- [Templates](https://github.com/coder/coder/blob/bbbe7aa2687671c8362ab3561c040d34ba0cca19/coderd/database/dump.sql#L1287-L1287)
- [Workspace Builds](https://github.com/coder/coder/blob/bbbe7aa2687671c8362ab3561c040d34ba0cca19/coderd/database/dump.sql#L1592-L1592)
- And now workspaces

If this is not done, the backend has to do multiple fetches when trying to retrieve information always displayed like the Owner's username. These fetches have the RBAC issue that they require the `READ` permission the related object, which **gets even more complicated with custom roles**. 

Without these joins, if you give `READ` permission to a workspace in an org, it also requires the permission to `READ` the user only to fetch the username & avatar. When using custom roles, these implicit links are confusing to use, and have caused a reported error.

We can further make this better, as some overfetching is still done, but this change is already a lot.

# What this costs

We have to maintain yet another view. _Insert rant about SQLc supporting client-side views that update with the schema_.

# RBAC Change

Giving `READ` to a workspace now lets you see all these fields without needing perms to the related objects. 

```sql
-- Owner
coalesce(visible_users.avatar_url, '') AS owner_avatar_url,
coalesce(visible_users.username, '') AS owner_username,
-- Organization
coalesce(organizations.name, '') AS organization_name,
coalesce(organizations.display_name, '') AS organization_display_name,
coalesce(organizations.icon, '') AS organization_icon,
coalesce(organizations.description, '') AS organization_description,
      -- Template
coalesce(templates.name, '') AS template_name,
coalesce(templates.display_name, '') AS template_display_name,
coalesce(templates.icon, '') AS template_icon,
coalesce(templates.description, '') AS template_description
```
# Guardrails

Converting to `WorkspaceTable` is guarded to stay in sync with the view. If you update the view, but not the converter function, this test will fail.

https://github.com/coder/coder/blob/bbbe7aa2687671c8362ab3561c040d34ba0cca19/coderd/database/modelqueries_internal_test.go#L19-L22.

And static structs are also guarded. Does the same as above, but ignores the converter function.

https://github.com/coder/coder/blob/bbbe7aa2687671c8362ab3561c040d34ba0cca19/coderd/database/gentest/models_test.go#L69-L69

We use `WorkspaceTable` for audit logs, since fields like `Username` are mutable, it would not make sense to include in the audit log with a workspace. It is sourced from the mutable field of Users.

### CODE REVIEW

File: docs/admin/security/audit-logs.md [hunk 1]
```diff
@@ -26,9 +26,9 @@ We track the following resources:
...
...td></tr><tr><td>name</td><td>true</td></tr><tr><td>organization_id</td><td>false</td></tr><tr><td>owner_id</td><td>true</td></tr><tr><td>template_id</td><td>true</td></tr><tr><td>ttl</td><td>true</td></tr><tr><td>updated_at</td><td>false</td></tr></tbody></table>                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |
```
Comment (johnstcn) [inline docs/admin/security/audit-logs.md:31]: Can we keep the old resource name? I don't think admins care if there is a difference between a change to a `WorkspaceTable` or a change to a `Workspace`?

**Review by johnstcn — COMMENTED** [review]
submitting partial review now

File: docs/admin/security/audit-logs.md (hunk 1 above)
Comment (Emyrk) [reply docs/admin/security/audit-logs.md:31]: The type is still `workspace` to admins: https://github.com/coder/coder/blob/b05b00c797506ed6a0512b9f164f7ee7d71fece5/coderd/audit/request.go#L184-L185

It is just `WorkspaceTable` in Go

File: coderd/database/modelqueries_internal_test.go [hunk 2]
```diff
@@ -13,3 +15,38 @@ func TestIsAuthorizedQuery(t *testing.T) {
 	_, err := insertAuthorizedFilter(query, "")
 	require.ErrorContains(t, err, "does not contain authorized replace string", "ensure replace string")
 }
+
+// TestWorkspaceTableConvert verifies all workspace fields are converted
+// when reducing a `Workspace` to a `WorkspaceTable`.
+// This test is a guard rail to prevent developer oversight mistakes.
+func TestWorkspaceTableConvert(t *testing.T) {
```
Comment (johnstcn) [inline coderd/database/modelqueries_internal_test.go:22]: praise: nice test!

File: testutil/reflect.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+func populateValue(v reflect.Value, r *Random) (reflect.Value, error) {
+	var err error
+
+	// Handle some special cases
+	switch v.Type() {
+	case reflect.TypeOf(time.Time{}):
+		v.Set(reflect.ValueOf(time.Date(2020, 5, 2, 5, 19, 21, 30, time.UTC)))
+		return v, nil
```
Comment (johnstcn) [inline testutil/reflect.go:74]: suggestion: should this be in `testutil.Random.Time` instead?

File: testutil/reflect.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+	return nil
+}
+
+func populateValue(v reflect.Value, r *Random) (reflect.Value, error) {
+	var err error
+
+	// Handle some special cases
+	switch v.Type() {
+	case reflect.TypeOf(time.Time{}):
+		v.Set(reflect.ValueOf(time.Date(2020, 5, 2, 5, 19, 21, 30, time.UTC)))
```
Comment (johnstcn) [inline testutil/reflect.go:73]: suggestion: use a non-UTC timezone

**Review by johnstcn — APPROVED** [review]

File: testutil/reflect.go (hunk 4 above)
Comment (Emyrk) [reply testutil/reflect.go:73]: I am going to be lazy here.

**Pull request merged**